draft-specifications/Locators.md

```diff
@@ -0,0 +1,137 @@
+  determined by the netlayer
+- **Transport**: A unique identifier to specify a netlayer
+- **Hints**: A hashmap of additional connection information.
+
+When comparing two node locators, the designator and transport are the only
+pieces of information which need to match. Two node locators can have the same
+designator and transport but  different hints and be considered to be the same
+node.
+
+## [Syrup Serialization](#node-syrup-serialization)
+
+It's encoded as a record with the label `ocapn-node` and three arguments:
+
+```
+<ocapn-node designator  ; string
+            transport   ; symbol (cannot contain ".")
+            hints>      ; hashmap | false
+```
+
+### Hints
+
+This is a hashmap of key and values which are designed to encode additional
+connection information that the  netlayer might need to reach the node. The keys
+should be symbols with the values as strings.
+
+There can be any number of hints, including none at all. If no hints are used
+this field should be set to false.
+
+## URI Serialization
+
+The URI serialization is a URI form designed to be provided to users to be given
+out of band to bootstrap a CapTP session.
+
+The URI format is as follows:
+```
+;; Without hints (i.e. hints are set to false)
+ocapn://<designator>.<transport>
+
+;; With hints
+ocapn://<designator>.<transport>?hint1=value1&hint2=value2
+```
+
+This is a URI with the scheme `ocapn` followed by the designator, a `.` and then
+the transport name. If any  hints exist they're added as part of the query
+parameters, otherwise emitted. 
+
+Note that the designator permits `.` to be used within it, however the final `.`
+should designate the seperator  between the designator segment and the transport
+identifier.
+
+# Sturdyref Locator
+
+A sturdyref locator includes a [Node Locator](#node-locator) and
+a `swiss-num` which represents a specific object located at that
+node. This should be considered a capability with this information
+alone being used to obtain a CapTP reference the given object.
+
+The pieces of information encoded in the sturdyref are:
+
+- [Node Locator](#node-locator)
+- Swiss number: string used to obtain an object
+
+## Syrup Serialization
+
+It's encoded as a record with the label `ocapn-sturdyref` and two arguments:
+
+```
+<ocapn-sturdyref node swiss-num>
+```
+
+The arguments are:
+
+- **node**: Syrup record defined in the [Syrup serialization of the Node locator](#node-syrup-serialization)
+- **swiss-num**: String which identifies the object.
+
+## URI Serialization
+
+The URI format follows a similar format to the node URI format, except with a
+"/s/" suffix to denoate that it's a sturdyref, followed by the swiss number
+value. Any hints are placed at the end of the URI.
+
+The URI format is as follows:
+```
+;; Without hints (i.e. hints are set to false)
+ocapn://<designator>.<transport>/s/<swiss-num>
+
+;; With hints
+ocapn://<designator>.<transport>/s/<swiss-num>?hint1=value1&hint2=value2
```

Comment:
Do the hints relate to the transport or to the ref?

If it is the latter, then the ref can be encoded as an URI `fragment`, leaving hierarchy part (authority+path) to the node:

`ocapn://<designator>.<transport>/?hint1=value1&hint2=value2#swiss-num`

Reply:
The hints are just additional information which might be needed to connect to the node. The onion netlayer doesn't need any but for example in Goblins there is a TCP+TLS netlayer which has the IP and port encoded in the hints (the designator is the hash of the TLS certificate).

One of the reasons `/s/swiss-num` was chosen was leaving the option to support many different locator types in the future (e.g. certificate locator types). This would then allow for `/c/...`  to be used and it'd be quite flexible.

Reply:
It is possible (desirable, even) to have a type prefix in the fragment too (`ocapn://.....#sn:swiss-num`).

What irks me in `node/s/path?node-hints` is that a reference is injected in a middle of a node information. This prevents using of a simple tail split and potentially clashes with path (or path-like) information used by a node (see "http-over-unix-sockets" encoding problems).

Reply:
Yeah this could be moved to the fragment section. Although I wonder if we should first decide in the group #74, if these things should be URIs at all. If they are I think this change could make sense.